### PR TITLE
Add support for soft delete related functions 

### LINKED ISSUE #64: add support for soft delete related functions
### Is your feature request related to a problem? Please describe.

For soft deleted records, DAL doesn't support finding records including them.

### Describe the solution you'd like

1. For DataStore and ProtoStore, add functions for finding records including soft deleted
2. For DataStore and ProtoStore, add functions for finding all records including soft deleted
3. return metaData in ProtoStore's soft delete function

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by YanhaoLi07**
- For DataStore and ProtoStore, add functions for finding records including soft deleted
- For DataStore and ProtoStore, add functions for finding all records including soft deleted
- Return metaData in ProtoStore's soft delete function

Fixes: https://github.com/vmware-labs/multi-tenant-persistence-for-saas/issues/64

### CODE REVIEW

File: pkg/datastore/database.go [hunk 1]
```diff
@@ -274,13 +295,22 @@ func (db *relationalDb) FindAllInTable(ctx context.Context, tableName string, re
 	return db.FindWithFilterInTable(ctx, tableName, record, records, pagination)
 }
 
+func (db *relationalDb) FindAllInTableIncludingSoftDeleted(ctx context.Context, tableName string, records interface{}, pagination *Pagination) error {
```
Comment (krishnamiriyala) [inline pkg/datastore/database.go:298]: FindAllInTableIncludingSoftDeletes?

File: pkg/datastore/datastore.go [hunk 2]
```diff
@@ -47,8 +47,10 @@ import (
 
 type DataStore interface {
 	Find(ctx context.Context, record Record) error
+	FindSoftDeleted(ctx context.Context, record Record) error
 	FindAll(ctx context.Context, records interface{}, pagination *Pagination) error
```
Comment (krishnamiriyala) [inline pkg/datastore/datastore.go:51]: replicate this too?

File: pkg/datastore/database.go [hunk 3]
```diff
@@ -216,16 +215,28 @@ func (db *relationalDb) Find(ctx context.Context, record Record) error {
 	return db.FindInTable(ctx, GetTableName(record), record)
 }
 
+func (db *relationalDb) FindSoftDeleted(ctx context.Context, record Record) error {
+	return db.FindSoftDeletedInTable(ctx, GetTableName(record), record)
+}
+
 func rollbackTx(tx *gorm.DB, db *relationalDb) {
```
Comment (krishnamiriyala) [inline pkg/datastore/database.go:222]: move to top

File: pkg/datastore/database.go [hunk 4]
```diff
@@ -54,8 +54,7 @@ import (
 const (
 	DbConfigOrgId      = "multitenant.orgId"      // Name of Postgres run-time config. parameter that will store current user's org. ID
 	DbConfigInstanceId = "multitenant.instanceId" // Name of Postgres run-time config. parameter that will store current session's instance ID
-
-	MaxIdleConns = 1
+	MaxIdleConns       = 1
```
Comment (krishnamiriyala) [inline pkg/datastore/database.go:57]: revert

File: pkg/protostore/protostore.go [hunk 5]
```diff
@@ -48,9 +48,11 @@ type ProtoStore interface {
 	Update(ctx context.Context, id string, msg proto.Message) (rowsAffected int64, md Metadata, err error)
 	Upsert(ctx context.Context, id string, msg proto.Message) (rowsAffected int64, md Metadata, err error)
 	FindById(ctx context.Context, id string, msg proto.Message, metadata *Metadata) error
+	FindByIdIncludingSoftDeleted(ctx context.Context, id string, msg proto.Message, metadata *Metadata) error
 	FindAll(ctx context.Context, msgs interface{}, pagination *datastore.Pagination) (metadataMap map[string]Metadata, err error)
+	FindAllIncludingSoftDeleted(ctx context.Context, msgs interface{}, pagination *datastore.Pagination) (metadataMap map[string]Metadata, err error)
 	FindAllAsMap(ctx context.Context, msgsMap interface{}, pagination *datastore.Pagination) (metadataMap map[string]Metadata, err error)
-	SoftDeleteById(ctx context.Context, id string, msg proto.Message) (rowsAffected int64, err error)
+	SoftDeleteById(ctx context.Context, id string, msg proto.Message) (rowsAffected int64, md Metadata, err error)
```
Comment (krishnamiriyala) [inline pkg/protostore/protostore.go:55]: thank you

File: pkg/protostore/protostore.go [hunk 6]
```diff
@@ -413,12 +440,69 @@ func (p ProtobufDataStore) FindAll(ctx context.Context, msgs interface{}, pagina
 	return metadataMap, nil
 }
 
-func (p ProtobufDataStore) SoftDeleteById(ctx context.Context, id string, msg proto.Message) (int64, error) {
+func (p ProtobufDataStore) FindAllIncludingSoftDeleted(ctx context.Context, msgs interface{}, pagination *datastore.Pagination) (metadataMap map[string]Metadata, err error) {
```
Comment (krishnamiriyala) [inline pkg/protostore/protostore.go:443]: Do you need replicate the whole method? or can create helper method?

**Review by krishnamiriyala — CHANGES_REQUESTED** [review]
nits: please add unit-tests, you can call softdelete and findwithSoftDeletes for one of the existing tests to cover all the new codepaths.

File: pkg/datastore/database.go (hunk 1 above)
Comment (YanhaoLi07) [reply pkg/datastore/database.go:298]: FindAllInTable() doesn't returns those that are already soft deleted. FindAllInTableIncludingSoftDeleted() includes those

File: pkg/datastore/datastore.go (hunk 2 above)
Comment (YanhaoLi07) [reply pkg/datastore/datastore.go:51]: Sorry didn't get it.

File: pkg/protostore/protostore.go (hunk 6 above)
Comment (YanhaoLi07) [reply pkg/protostore/protostore.go:443]: Replication for backward compatibility. Or add a "softDelete" flag directly inside FindAllIncludingSoftDeleted() but that modifies function signatures for all usages

Comment (YanhaoLi07) [reply pkg/protostore/protostore.go]: change pushed. Please check latest

File: pkg/protostore/protostore.go (hunk 5 above)
Comment (YanhaoLi07) [reply pkg/protostore/protostore.go:55]: Hi. Do I need to make any changes in this function?

File: pkg/datastore/datastore.go (hunk 2 above)
Comment (YanhaoLi07) [reply pkg/datastore/datastore.go:51]: Could you please explain more? Do you mean I shouldn't replicate this function?

File: pkg/protostore/protostore.go [hunk 7]
```diff
@@ -413,12 +443,32 @@ func (p ProtobufDataStore) FindAll(ctx context.Context, msgs interface{}, pagina
 	return metadataMap, nil
 }
 
-func (p ProtobufDataStore) SoftDeleteById(ctx context.Context, id string, msg proto.Message) (int64, error) {
+// FindAll Finds all messages (of the same type as the element of msgs) in Protostore and stores the result in msgs.
+// msgs must be a pointer to a slice of Protobuf structs or a pointer to a slice of pointers to Protobuf structs.
+// It will be modified in-place.
+// Returns a map of Protobuf messages' IDs to their metadata (parent ID & revision).
+func (p ProtobufDataStore) FindAll(ctx context.Context, msgs interface{}, pagination *datastore.Pagination) (metadataMap map[string]Metadata, err error) {
+	return p.findAll(ctx, msgs, pagination, false)
```
Comment (YanhaoLi07) [inline pkg/protostore/protostore.go:451]: use findAll() helper function to reuse common code between FindAll() and FindAllIncludingSoftDeleted()

File: pkg/datastore/database.go (hunk 3 above)
Comment (YanhaoLi07) [reply pkg/datastore/database.go:233]: please check latest commit. Deleted unused functions

File: pkg/datastore/database.go (hunk 1 above)
Comment (krishnamiriyala) [reply pkg/datastore/database.go:307]: please have common method as the implementation is same  findAllInTable() with softDelete as param.

File: pkg/datastore/datastore.go (hunk 2 above)
Comment (krishnamiriyala) [reply pkg/datastore/datastore.go:51]: FindAllSoftDeleted

File: pkg/datastore/database.go (hunk 3 above)
Comment (krishnamiriyala) [reply pkg/datastore/database.go:233]: nit: just have all related functions like Find/FindSoftDeleted/FindInTable all close to each other and rollbackTx is common can be moved closer to first method definition.

File: pkg/protostore/protostore.go (hunk 5 above)
Comment (krishnamiriyala) [reply pkg/protostore/protostore.go:55]: nope

**Review by krishnamiriyala — COMMENTED** [review]
overall lgtm. lmk once you add unit-tests

File: pkg/datastore/database.go [hunk 8]
```diff
@@ -240,25 +254,37 @@ func (db *relationalDb) FindInTable(ctx context.Context, tableName string, recor
 		}
 	}()
 
-	if err = tx.Table(tableName).Where(record).First(record).Error; err != nil {
-		if errors.Is(err, gorm.ErrRecordNotFound) {
-			return ErrRecordNotFound.Wrap(err).WithValue("record", fmt.Sprintf("%+v", record)).WithValue(DB_NAME, db.dbName)
+	if softDelete {
```
Comment (krishnamiriyala) [inline pkg/datastore/database.go:257]: you could just use softDelete for setting tx := tx.Unscoped()

File: pkg/datastore/database.go [hunk 9]
```diff
@@ -297,28 +344,38 @@ func (db *relationalDb) FindWithFilterInTable(ctx context.Context, tableName str
...
 		}
 	}
-	if err = tx.Find(records).Error; err != nil {
-		db.logger.Debug(err)
-		return ErrExecutingSqlStmt.Wrap(err).WithValue(DB_NAME, db.dbName)
-	}
 
-	if !db.txFetcher.IsTransactionCtx(ctx) {
-		if err = tx.Commit().Error; err != nil {
+	if softDelete {
```
Comment (krishnamiriyala) [inline pkg/datastore/database.go:366]: same here

File: pkg/datastore/datastore_test.go [hunk 10]
```diff
@@ -164,7 +167,19 @@ func testCrud(t *testing.T, ds datastore.DataStore, ctx context.Context, myCokeA
...
+	for _, record := range []*AppUser{user1, user2} {
+		rowsAffected, err := ds.SoftDelete(ctx, record)
+		assert.NoError(err)
+		assert.EqualValues(1, rowsAffected)
+		queryResult := &AppUser{Id: record.Id}
+		err = ds.Find(ctx, queryResult)
+		assert.ErrorIs(err, ErrRecordNotFound)
+		assert.True(queryResult.AreNonKeyFieldsEmpty())
+		err = ds.FindSoftDeleted(ctx, queryResult)
+		assert.NoError(err)
```
Comment (krishnamiriyala) [inline pkg/datastore/datastore_test.go:182]: assert with cmp.Equal for result

File: pkg/datastore/datastore_with_txcontext_test.go [hunk 11]
```diff
@@ -189,7 +191,7 @@ func TestSingleTxCrud(t *testing.T) {
 		queryResult := &AppUser{Id: record.Id}
 		err = ds.Find(txCtx, queryResult)
 		assert.NoError(err)
-		assert.Equal(record, queryResult)
+		cmp.Equal(record, &queryResult, cmpopts.IgnoreFields(AppUser{}, "CreatedAt", "UpdatedAt", "DeletedAt"))
```
Comment (krishnamiriyala) [inline pkg/datastore/datastore_with_txcontext_test.go:194]: nit: can be taken up later, helper method that prints the records too? Does it assert?

File: pkg/datastore/datastore_with_txcontext_test.go [hunk 12]
```diff
@@ -83,7 +85,7 @@ func testTxCrud(t *testing.T, ds datastore.DataStore, ctx context.Context, user1
...
```
Comment (krishnamiriyala) [inline pkg/datastore/datastore_with_txcontext_test.go:88]: This doesn't seem to assert only return bool

File: pkg/protostore/protostore.go [hunk 13]
```diff
@@ -275,27 +278,55 @@ func (p ProtobufDataStore) UpsertWithMetadata(ctx context.Context, id string, ms
...
+func (p ProtobufDataStore) findById(ctx context.Context, id string, msg proto.Message, metadata *Metadata, softDelete bool) error {
 	protoStoreMsg, err := p.MsgToFilter(ctx, id, msg)
 	if err != nil {
 		return err
 	}
 
-	err = p.ds.Find(ctx, protoStoreMsg)
-	if err != nil {
-		return err
+	if softDelete {
```
Comment (krishnamiriyala) [inline pkg/protostore/protostore.go:294]: see if you can push all if/elses to datastore and just pass bool

File: test/data.go [hunk 14]
```diff
@@ -85,9 +90,12 @@ func (a AppUserSlice) Less(x, y int) bool {
 func (a AppUserSlice) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
 
 type App struct {
-	Id       string `gorm:"primaryKey;column:application_id"`
-	Name     string
-	TenantId string `gorm:"primaryKey;column:org_id"`
+	Id        string `gorm:"primaryKey;column:application_id"`
+	Name      string
+	TenantId  string `gorm:"primaryKey;column:org_id"`
+	CreatedAt time.Time
```
Comment (krishnamiriyala) [inline test/data.go:96]: use gorm.Model instead similar to protostoreMsg

File: pkg/datastore/database.go (hunk 8 above)
Comment (YanhaoLi07) [reply pkg/datastore/database.go:257]: If tx is carried inside ctx, tx := tx.Unscoped() will make all following functions calls to have tx.Unscoped(), which creates errors

File: test/data.go (hunk 14 above)
Comment (YanhaoLi07) [reply test/data.go:96]: gorm.Model itself has "Id" as primary key, which has conflict with parent struct's Id field.

File: pkg/datastore/database.go (hunk 8 above)
Comment (krishnamiriyala) [reply pkg/datastore/database.go:257]: tx can be a new variable too you are don't need to override

File: pkg/datastore/datastore_with_txcontext_test.go (hunk 12 above)
Comment (krishnamiriyala) [reply pkg/datastore/datastore_with_txcontext_test.go]: Is it easy to figure out the diff on failures or do you need debugger with assert.True?

File: pkg/protostore/protostore.go (hunk 13 above)
Comment (krishnamiriyala) [reply pkg/protostore/protostore.go]: Thank you

File: test/data.go (hunk 14 above)
Comment (krishnamiriyala) [reply test/data.go:96]: aah ok

File: pkg/datastore/database.go [hunk 15]
```diff
@@ -225,10 +228,21 @@ func rollbackTx(tx *gorm.DB, db *relationalDb) {
 	}
 }
 
+// Find Finds a single record that has the same values as non-zero fields in the record.
```
Comment (krishnamiriyala) [inline pkg/datastore/database.go:231]: I dont see DOCUMENTATION.md change, can you run `make doc`

File: pkg/datastore/datastore_test.go [hunk 16]
```diff
@@ -21,6 +21,10 @@ package datastore_test
 import (
 	"context"
 	"fmt"
+	"github.com/bxcodec/faker/v4"
```
Comment (krishnamiriyala) [inline pkg/datastore/datastore_test.go:24]: `make lint_fix`

**Review by krishnamiriyala — CHANGES_REQUESTED** [review]
Please run `make lint_fix doc`, other comments can be in follow up

**Review by krishnamiriyala — APPROVED** [review]

File: pkg/datastore/database.go (hunk 8 above)
Comment (krishnamiriyala) [reply pkg/datastore/database.go:257]: @YanhaoLi07 resolving for now to unblock Gaurav, please follow up with a change

**Review by krishnamiriyala — APPROVED** [review]

**Pull request merged**